enhancements/cluster-runtime-constraints.md

```diff
@@ -0,0 +1,138 @@
+---
+title: Cluster Runtime Constraint
+authors:
+  - "@dinhxuanvu"
+reviewers:
+  - "@kevinrizza"
+approvers:
+  - TBD
+creation-date: 2021-08-26
+last-updated: 2020-09-06
+status: provisional
+---
+
+# cluster-runtime-constraint
+
+## Release Signoff Checklist
+
+- [ ] Enhancement is `implementable`
+- [ ] Design details are appropriately documented from clear requirements
+- [ ] Test plan is defined
+- [ ] Graduation criteria for dev preview, tech preview, GA
+
+## Summary
+
+This proposal presents the specification and usage to enable runtime constraints to be considered during operator dependency resolution. This feature will ensure Operator Lifecycle Manage (OLM) to install operators that satisfy dependency requirements and runtime constraints if present.
+
+## Motivation
```

Comment:
1. The specific constraint that will be used first is kubernetes version (on Red Hat's term OpenShift version).
2. We prefer a solution that is generic enough that it can be extendable in the future without overhauling the entire feature.
3. As I answered in the other comment, auto-detection is not addressed at this time but I expect it to be addressed later for sure.

Reply:
Thanks for elaborating. It would be good to add some of this to the proposal.

Most package managers pre-define certain pieces of information that can be used to determine whether a package can be installed. That includes comparing versions in some standard way, checking system architecture, checking for other capabilities declared in some standard way (for example an RPM might need a compatible shell but doesn't care which one), etc. It seems worth at least seriously considering whether OLM could make available a reasonable list of observed cluster properties and enable packagers to match against them. It might not be a long list, and could be a LOT more usable especially from the cluster admin perspective.

Reply:
> It seems worth at least seriously considering whether OLM could make available a reasonable list of observed cluster properties and enable packagers to match against them. It might not be a long list, and could be a LOT more usable especially from the cluster admin perspective.

Maybe I'm splitting hairs, but it seems like this design leaves the possibility open for k8s distributions that package OLM to decide what cluster properties make sense to provision.

- OpenShift might expose a property called "ocpVersion"
- GKE might expose a property called "gkeVersion"

I _do_ think it could make sense to consider what it would look like to make a set of generic cluster properties available on all clusters. Otherwise this feature doesn't seem very valuable to upstream bundle authors.

Reply:
> The specific constraint that will be used first is kubernetes version

Are there other package managers that actually encode platform version in constraints? If so it might be helpful to call those out as examples. I'm most familiar with debian/rpm repositories which are usually curated and versioned such that a particular OS version pulls from an OS version-specific repository, and there no OS-level mechanism to prevent a user from installing packages from outside that repository.

Reply:
I am late to the battle but I really agree with what @mhrivnak wrote. A good example of such a mechanism is the [node feature discovery](https://github.com/kubernetes-sigs/node-feature-discovery), which retrieves hardware and system information from nodes and make them consumable by others. There is then the [special resource operator](https://github.com/openshift/special-resource-operator), NVIDIA operator and others that consume this information for installing driver modules on a subset of nodes for instance.
Most of the operator writers are software providers they need to rely upfront on a clear set of properties. They cannot know of each individual cluster their operator may be installed on and requiring cluster-admins to define these properties removes the benefit of having this automation